### PR TITLE
push JSON Path evaluation down to storage layer

### PR DESCRIPTION
**Pull request by richardstartin**
Pushes JSON path evaluation down to the storage layer (giving direct access to dictionaries and forward index) which avoids various intermediate materialisations of strings, byte arrays and so on. The benefit to users is the potential to avoid a lot of allocation of large `byte[]` and `String` once the JsonPath library can accept UTF-8 encoded `byte[]`. This also creates an SPI to make the evaluation logic pluggable. The same pushdown mechanism could be abstracted to make extensible to regular expressions etc. in the future.

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/common/DataBlockCache.java
Comment (mayankshriv) [inline pinot-core/src/main/java/org/apache/pinot/core/common/DataBlockCache.java:131]: Does it make sense to have a more generic evaluator abstraction (with JsonPathEvaluator as one of the impls) here?

Comment (richardstartin) [reply pinot-core/src/main/java/org/apache/pinot/core/common/DataBlockCache.java:131]: I believe in the rule of 3: create a generic abstraction once you have three implementations.

Comment (richardstartin) [reply pinot-core/src/main/java/org/apache/pinot/core/common/DataBlockCache.java:131]: There is no external impact of changing `JsonPathEvaluator` to `Evaluator` which could be a regular expression (or something else) later, it all stays in pinot-core.

**Comment by richardstartin** [comment]
The try/catches can be removed if https://github.com/json-path/JsonPath/pull/767 is merged and released.

File: pinot-core/src/main/java/org/apache/pinot/core/common/DataBlockCache.java
Comment (kkrugler) [inline pinot-core/src/main/java/org/apache/pinot/core/common/DataBlockCache.java:317]: Isn't this getting values for a multi-valued column? Same comment for Javadocs below.

Comment (richardstartin) [reply pinot-core/src/main/java/org/apache/pinot/core/common/DataBlockCache.java:317]: No: the JSON column is single value, but the JsonPath produces a multi-valued result. If it’s confusing the verbiage can be expanded upon.

File: pinot-core/src/main/java/org/apache/pinot/core/common/DataBlockCache.java
Comment (mayankshriv) [reply pinot-core/src/main/java/org/apache/pinot/core/common/DataBlockCache.java:131]: Yeah I see your point about the rule of 3. I think I put the comment in a less effective place, since this is not an interface implementation.

Comment (richardstartin) [reply pinot-core/src/main/java/org/apache/pinot/core/common/DataBlockCache.java:131]: OK. I think what’s important about this API is the _shape_ of the result, and JsonPath evaluation will only ever produce a single vector of values with the same row indexing, in the same way that, say, a pushed down regular expression would. There are other obvious shapes to consider, such as an aggregator which produces a single value, or potentially something like a JSON UNNEST function which produces a matrix of values, typically with different row indexing. So any early abstraction for this needs to be able to model shape adequately. JsonPathEvaluator could be generalised to _TransformEvaluator_ because it preserves the row indexing.

**Comment by kishoreg** [comment]
I like the concept of the pushdown evaluator. Few thoughts/concerns
- Some methods added to projectionBlock seem to be specific to projectionBlock and is deviating from the block concept of chaining any blocks together i.e. the caller of the block does not understand the implementation of the underlying block. We have violated this in few places and we should fix that. It might be a good idea to enhance the Block interface to allow a pushdown evaluator.
- Conceptually, we are moving from a Column to Column + path but modelling the path as an evaluator. But the query layer is deciding the evaluator instead of the storage layer. 
- The evaluator seems to be set at the JVM level but in practice, it's possible for each table to have a different evaluator and in fact, it's possible that each segment can have a different evaluator.

At a high level, the segment is the smallest unit of query processing in Pinot and most of the things can be decided independently at that scope - segmentDirectory, readers, writers, Blocks, indexes, encoding etc.

Having said that, I don't have a very good solution yet but I think we can come up with something after a couple of iterations. One thought is enhancing the concept (String Column) to a Column Object which has  SingleColumn, NestedColumn( column + path) as implementations.

Overall, I think the Storage (Reader) should decide on the evaluator instead of the query layer.

**Comment by richardstartin** [comment]
> Conceptually, we are moving from a Column to Column + path but modelling the path as an evaluator. But the query layer is deciding the evaluator instead of the storage layer.
>
> The evaluator seems to be set at the JVM level but in practice, it's possible for each table to have a different evaluator and in fact, it's possible that each segment can have a different evaluator.

The query layer does not decide, the query layer just _produces_ the evaluator from the query. It does not know what evaluator implementation it is producing, just whatever has been registered. It just won’t push the evaluator down unless the identifier is a column identifier rather than an expression.

The evaluator, having captured information from the query, is pushed down to the storage layer. The storage layer then presents storage layer capabilities (I.e. dictionary + forward index) to the evaluator, which can interrogate and choose what to do with them. 

Think of the evaluator as a _mediator_ between the query and storage layers for which any implementation can be registered via SPI. It encapsulates the logic of the function, and needs to know all relevant storage capabilities present in the deployment. The right evaluator to register for a given Pinot _deployment_ is the one that matches all storage capabilities in the deployment.

The evaluator must decide how to evaluate its function based on what storage capabilities it is presented with in evaluateBlock. If evaluateBlock is presented with, say, a BytesDictionary, it must be able handle extracting JSON encoded as bytes from the dictionary and evaluating a JSON path against it, just as it must be able to handle reading JSON from a raw forward index, or from whatever ForwardIndexReader from whatever storage plugin it is presented with. 

To illustrate the evaluator’s responsibility to understand the storage, the DefaultJsonPathEvaluator in this PR knows how to work with all storage capabilities capable of storing JSON currently present in Pinot (dictionarised string, bytes, raw string, bytes, JSON type column or not) but adding a storage plugin means creating a new evaluator  which understands the custom storage but delegating to a default evaluator for everything else, and registering it via the SPI.

This has the ergonomic benefit of being able to implement storage plugins which deviate from the ForwardIndexReader interface. Completely new access patterns and the necessary interfaces can be implemented without changing the ForwardIndexReader interface: 
1. Implement new ForwardIndexReader with a previously inconceivable API
2. Implement an evaluator which uses the new API for instances of that ForwardIndexReader, otherwise delegate to default implementation 
3. Register the evaluator 
4. (Optional) later enhance the ForwardIndexReader interface with method supporting a tried and proven access pattern, explain to everybody why it’s a good idea with data to refer to

This circumnavigates the need to form design consensus before having a proven storage implementation and should reduce the number of changes to the ForwardIndexReader interface to those which have already been **proven to make sense**, while never touching the query layer once the relevant TransformFunction has been modified to push an evaluator down.

**Comment by richardstartin** [comment]
> Conceptually, we are moving from a Column to Column + path but modelling the path as an evaluator. But the query layer is deciding the evaluator instead of the storage layer.

Specifically on this, not really, the model has the potential to be far more generic than that. It’s more going from column to column + parameters + arbitrary sequence of storage layer API calls. 

This is good, because within the context of an evaluator, the most efficient API calls can be made. Case in point: in DefaultJsonPathEvaluator it’s possible to avoid converting byte[] to String. It would also be possible to use hypothetical new zero copy (but unsafe) methods to read bytes on the basis that we don’t need to create a vector of JSON records, but a vector of JSON evaluation results. Individual evaluator implementations can be arbitrarily efficient (and complex) in a way that prevents the query layer from ever seeing any of this complexity, keeping the query layer clean just as it avoids trying to model all of the complex access patterns we might like in the storage layer with a single interface.

File: pinot-core/src/main/java/org/apache/pinot/core/common/DataBlockCache.java
Comment (richardstartin) [reply pinot-core/src/main/java/org/apache/pinot/core/common/DataBlockCache.java]: I just pushed something which should address your concerns.

**Comment by richardstartin** [comment]
@kishoreg please see the updated Javadoc which provides more colour.

@siddharthteotia you expressed some interest in this PR, would you mind taking a look please?

**Comment by richardstartin** [comment]
> Some methods added to projectionBlock seem to be specific to projectionBlock and is deviating from the block concept of chaining any blocks together i.e. the caller of the block does not understand the implementation of the underlying block. We have violated this in few places and we should fix that. It might be a good idea to enhance the Block interface to allow a pushdown evaluator.

@kishoreg this has been deviated from significantly already, and in many `Block` implementations there would be no sensible implementation but to throw `UnsupportedOperationException` which breaks LSP anyway. 

I also want to call out that there appears to be an element of complexity which you may have missed: `TransformFunction` isn't guaranteed to operate on directly on storage, it may also operate on the output of some other JSON function (I added a test for this) - when that happens, there needs to be an implementation which can operate on values produced by another `TransformFunction`. This (to push down or evaluate directly) is a decision which can only be made in the query layer. I'm not sure if you didn't mention this because it makes sense already or not.

File: pinot-core/src/main/java/org/apache/pinot/core/common/DataFetcher.java
Comment (klsince) [inline pinot-core/src/main/java/org/apache/pinot/core/common/DataFetcher.java:105]: nit: is `JSON column` too specific for the method?

File: pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/TransformBlock.java
Comment (klsince) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/TransformBlock.java:85]: any need to check `if (expression.getType() == ExpressionContext.Type.IDENTIFIER)` like L51 above? iiuc, we pushDown only when there is no transforms, i.e. a pass-through to projection block.

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/evaluator/TransformEvaluator.java
Comment (klsince) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/evaluator/TransformEvaluator.java:28]: nit: `JSON path` is too specific here?

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/evaluator/json/JsonPathEvaluators.java
Comment (klsince) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/evaluator/json/JsonPathEvaluators.java:95]: 👍

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/PushDownTransformFunction.java
Comment (klsince) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/PushDownTransformFunction.java:25]: nit: maybe add some comments about who can implement this interface?

File: pinot-core/src/main/java/org/apache/pinot/core/common/DataFetcher.java
Comment (richardstartin) [reply pinot-core/src/main/java/org/apache/pinot/core/common/DataFetcher.java:105]: This is a carry over from before generalising the interface. Will fix all of them.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/TransformBlock.java
Comment (richardstartin) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/TransformBlock.java:85]: If there is a column name it must be an identifier. I had previously only implemented this for `ProjectionBlock` for this reason though.

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/evaluator/TransformEvaluator.java
Comment (richardstartin) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/evaluator/TransformEvaluator.java:28]: As above.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/PushDownTransformFunction.java
Comment (richardstartin) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/PushDownTransformFunction.java:25]: 👍🏻

**Review by amrishlal — CHANGES_REQUESTED** [review]

File: pinot-core/src/test/java/org/apache/pinot/queries/JsonPathQueriesTest.java
Comment (richardstartin) [reply pinot-core/src/test/java/org/apache/pinot/queries/JsonPathQueriesTest.java:76]: I added these because I think the test setup is very good, it can actually be used to test cases on `jsonextractscalar` that weren't previously tested, and these columns are only used to test `jsonextractscalar`. Indeed, JSON literals did not work and still do not work with `BYTES` and `STRING` Does this alleviate your concern?

File: pinot-core/src/test/java/org/apache/pinot/queries/JsonPathQueriesTest.java
Comment (richardstartin) [reply pinot-core/src/test/java/org/apache/pinot/queries/JsonPathQueriesTest.java:275]: I'm not sure I agree about that, I don't think that this test was stable w.r.t. extension of the test setup. 

In order to add columns to the test setup (e.g. to test a RAW JSON column, which was not previously tested, and is now) `expecteds1` would always need to be extended to include every column added. So there is a choice: add the values to `expected1`, duplicate the whole test _setup_ to add the new columns for a new test, or control what is projected to keep this test stable. Since this is a test for selection and not projection, I opted to do the latter. 

SO do you want me to duplicate the test setup so I can add tests for querying a RAW JSON column, or change the query back to `select *` and add all new columns to `expected1`?

Comment (richardstartin) [reply pinot-core/src/test/java/org/apache/pinot/queries/JsonPathQueriesTest.java]: I have made a change to prevent changing this particular test here (which any kind of schema change always will, the way the test has been written): https://github.com/apache/pinot/pull/7820/commits/76f703540e0f617799d7bce459c4c1fce3da5e85

However, my preference is to drop that commit and make the change above to make the test stable to schema changes. The schema change is beneficial, because it tests storage configurations (e.g. RAW JSON) which wasn't previously tested.

File: pinot-core/src/test/java/org/apache/pinot/queries/JsonPathQueriesTest.java
Comment (kishoreg) [reply pinot-core/src/test/java/org/apache/pinot/queries/JsonPathQueriesTest.java]: Unrelated to this PR. @amrishlal when did we make the decision that json_extract should work only on JSON datatype? this is a backward-incompatible change.

Comment (richardstartin) [reply pinot-core/src/test/java/org/apache/pinot/queries/JsonPathQueriesTest.java]: @kishoreg I don't think that's what @amrishlal was saying, this test focuses on jsonpath literals like `select * from table where foo.bar = 0` and these only work on JSON columns, so adding these other columns here so `jsonextractscalar` can be tested using the thorough set up this tests provides doesn't feel appropriate.

**Comment by richardstartin** [comment]
@kishoreg Specifying in Javadoc that these are unstable APIs is an excellent compromise. I think it should be possible to evolve towards what you want. I must admit that I don't see it right now, but I'm trying hard to find a way both to implement the suggestion and am thinking about the benefits. I want to make it very easy to implement storage plugins (as an idea incubation mechanism) without needing community consensus on the shape of SPIs like `ForwardIndexReader` which will affect a lot of stakeholders. A delegation chain of `TransformEvaluator`s, marked as an unstable interface, feels like a good way to iterate without stalling for consensus or imposing churn on stakeholders.

File: pinot-core/src/test/java/org/apache/pinot/queries/JsonPathQueriesTest.java
Comment (amrishlal) [reply pinot-core/src/test/java/org/apache/pinot/queries/JsonPathQueriesTest.java]: @richardstartin This unit test file is only to test JSON path expressions on `JSON` column. If you are trying to validate JSON_EXTRACT_SCALAR function, I think a better place for it would be JsonDataTypeQueriesTest.java which has a lot of test cases for json_extract_scalar. I haven't looked through the PR in detail, so maybe I am missing something, but will look into more detail.

@kishoreg Perhaps that was just my own personal impression at the time we were code reviewing JSON datatype. I suppose it should be ok to allow `JSON_EXTRACT_SCALAR` to work over `BYTE`, but then there is no guarantee that `BYTE` or `STRING` would contain JSON data (?), i.e any valid or invalid JSON data could be put into a `STRING` or `BYTE` column. If those byes or string are ingested into `JSON` datatype column, then we know for sure that it is valid JSON and can treat it as such during query processing, indexing, etc.

Comment (richardstartin) [reply pinot-core/src/test/java/org/apache/pinot/queries/JsonPathQueriesTest.java]: @amrishlal I split the tests so this test only tests JSON path expressions and not the extract scalar function, but moved the setup into a base class so it can be reused. Does this resolve this issue?

Comment (amrishlal) [reply pinot-core/src/test/java/org/apache/pinot/queries/JsonPathQueriesTest.java]: @richardstartin Yes, it resolves the issue. Thanks.

**Review by Jackie-Jiang — COMMENTED** [review]
Currently the data fetch has the buffers maintained within the `BlockValSet`. The benefit of doing so is that buffer for the same expression can be shared across different functions (through `DataBlockCache`). The disadvantage is that we need to keep the buffers for all unique expressions in memory, and map lookup is required to read the shared buffer.

In this PR, the buffer is maintained within the transform function, and directly pushed down to the forward index level to fill the values. It can be more efficient as the values are directly filled into the buffer, but values for the same expression in different functions need to be read multiple times (cache is bypassed).

Both ways have their own pros and cons, but we should try to stick with on of them, or it can be quite confusing, and we have to maintain 2 sets of different read apis.

To make the evaluator push down using the first approach, we can potentially extend `public BlockValSet getBlockValueSet(String column)` to `public BlockValSet getBlockValueSet(String column, @Nullable TransformEvaluator evaluator)`, and embed the evaluator into the `DataBlockCache`.

Or if we found the first approach is adding unnecessary overhead, we can remove the cache and always push down the buffer.

IMO I can see several use cases benefitting from the cache, and moving to the second approach might cause regression.

**Comment by richardstartin** [comment]
@Jackie-Jiang the reason I bypassed the cache is because the cache would need a composite key to avoid different JSON path transformations sharing results spuriously. It’s not a tangential optimisation to get rid of a few buffers or array copies. Could this be handled differently without the cache by reordering the execution graph to deduplicate common subexpressions? From my perspective, it doesn’t feel like a huge departure to allow some expressions to bypass the cache, and I don’t think it should be necessary for all expressions to bypass the cache if some do.

**Comment by richardstartin** [comment]
@Jackie-Jiang to demonstrate an open mind, I've implemented your suggestion here: https://github.com/apache/pinot/pull/7820/commits/f75135887001394abb4097a9ddae486e5cea1c48

I'm not convinced this is necessarily cleaner (`ProjectionBlock` is though) and I think it was better to keep the "tunnel" for evaluators separate from the regular path with caching. Writing the array copies from the enormous array maintained by `DataBlockCache` (because it doesn't have enough information to size the result properly) to the correctly sized arrays in `JsonExtractScalarFunction` made me want to cry ever so slightly. 

Please pay attention to the changes to the cache key in `DataBlockCache` - JSONPath evaluations need to be discriminated by both the JSON Path itself and the default value, on top of the column name and type of the result. This is probably better than having large JSON documents lingering in the cache, but I'm struggling to see this cache as an unequivocal benefit for every possible transformation. 

Could we avoid caching JSON path evaluations at the block level and revert the commit? If we find that we spend a lot of time on duplicate JSON path evaluations, could we think about avoiding the planning of duplicate evaluations instead?

**Comment by Jackie-Jiang** [comment]
> Could this be handled differently without the cache by reordering the execution graph to deduplicate common subexpressions?

I agree. Ideally we should deduplicate the expression during the query planning, and have only one transform function for each unique expression. Then we can remove the cache and avoid all the unnecessary overhead.

Towards this direction, I agree we should maintain the buffer within the transform layer and push the buffer down through the data fetcher. We should think of a smooth way to change the interface and move to the end state where all the transform functions (with or without an evaluator) can pass down the buffer to fill the values.

For now, we might just add some read methods into the `ProjectionBlock` so that it can be used by the `TransformFunction`? Adding these methods to the `Block` interface might not match our end goal.

Do you see a path how we can move to the end state without making temporary changes to the interfaces?

**Comment by richardstartin** [comment]
> For now, we might just add some read methods into the ProjectionBlock so that it can be used by the TransformFunction? Adding these methods to the Block interface might not match our end goal.

This is what the PR originally did, I did not modify the `Block` interface and added methods to the concrete `ProjectionBlock`, but it was requested above to ensure substitutability of different block types. It feels like I am violating an aspirational design principle with this PR, but as far as I can tell, it has been violated for a long time. A lot of good things will follow on from the ability to tunnel down to the storage, so I would like to find a way to progress here.

**Comment by richardstartin** [comment]
@Jackie-Jiang please see https://github.com/apache/pinot/pull/7820/commits/83dc232fc5d89dc1331437bc444512bfcd519eec - it removes `pushDown` from the `Block` interface.

**Comment by yupeng9** [comment]
Curious, do you have some benchmark number to show how much perf improvement?

**Comment by richardstartin** [comment]
> Curious, do you have some benchmark number to show how much perf improvement?

Performance improvement isn't the primary motivation but to provide a plugin mechanism which allows query logic to depend on the available storage. Before merging I would, of course, provide evidence that this does not _regress_ performance.

**Comment by kishoreg** [comment]
@Jackie-Jiang let's try to get to the final version in a few iterations. I think it's better to keep the changes restricted to projection block for now. Let's get this to work and we can try to revisit the block interface in a different PR.

I am +1 on this PR assuming we will mark the methods in projectionblock as evolving

**Review by kishoreg — APPROVED** [review]

**Comment by richardstartin** [comment]
I'll update the Javadoc but I told @siddharthteotia that this would not be merged until he's had a chance to review it.

File: pinot-core/src/main/java/org/apache/pinot/core/common/DataBlockCache.java
Comment (richardstartin) [reply pinot-core/src/main/java/org/apache/pinot/core/common/DataBlockCache.java:317]: Just circling back here - this is ambiguous now the evaluator concept is more general than JSON, so I will remove the comment about the arity of the column.

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/evaluator/json/JsonPathEvaluators.java
Comment (Jackie-Jiang) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/evaluator/json/JsonPathEvaluators.java:137]: Don't fully understand the logic here. What is the delegate evaluator used for?

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/evaluator/json/JsonPathEvaluator.java
Comment (Jackie-Jiang) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/evaluator/json/JsonPathEvaluator.java:29]: These interfaces do not belong to the `segment-spi`, but `query-spi` which is not available yet. For now we can put them into the `pinot-core` which contains all the query related interfaces

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/evaluator/json/JsonPathEvaluators.java
Comment (Jackie-Jiang) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/evaluator/json/JsonPathEvaluators.java:67]: Does it work if we simply put the `provider` as a `private static` field and set it during the instance startup? Several classes handle the plugin that way

**Review by Jackie-Jiang — COMMENTED** [review]
Mostly good. I think this is the cleanest solution without changing the existing data fetch mechanisms

File: pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/ProjectionBlock.java
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/ProjectionBlock.java:81]: +1

**Comment by siddharthteotia** [comment]
> I'll update the Javadoc but I told @siddharthteotia that this would not be merged until he's had a chance to review it.

@richardstartin, thanks for waiting. I pulled the changes locally and understood them. Had a few minor questions which I met with @Jackie-Jiang to clarify. LGTM

**Review by siddharthteotia — APPROVED** [review]
@richardstartin thanks for waiting

File: pinot-core/src/main/java/org/apache/pinot/core/common/DataFetcher.java
Comment (richardstartin) [reply pinot-core/src/main/java/org/apache/pinot/core/common/DataFetcher.java:754]: good catch, this was an unintentional change made during the churn

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/JsonExtractScalarTransformFunction.java
Comment (richardstartin) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/JsonExtractScalarTransformFunction.java:98]: This is intentionally like this. A plugin may not be implemented in terms of the JsonPath library, and even if it is, the compilation was cached across all segments in #7826

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/evaluator/json/JsonPathEvaluators.java
Comment (richardstartin) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/evaluator/json/JsonPathEvaluators.java:67]: I think this is better for a couple of reasons
1. double registration becomes obvious
2. various runtime benefits of the provider being `final`

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/evaluator/json/JsonPathEvaluators.java
Comment (richardstartin) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/evaluator/json/JsonPathEvaluators.java:137]: The idea here is there is a chain of responsibility. There is the standard way of evaluating JSON paths (the result of the refactoring here), but a plugin could be registered which adds special storage as well as an evaluator which can interpret that storage. There could be a mix of segments using the default storage and the plugin storage. On a case by case basis, the registered plugin either handles the encountered storage or delegates if it doesn't know how. 

I'm not sure what I was thinking when I wrote this particular line though - `DefaultJsonPathEvaluator` is the end of the chain and will never delegate, so its delegate will always be null.

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/evaluator/json/JsonPathEvaluator.java
Comment (richardstartin) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/evaluator/json/JsonPathEvaluator.java:29]: Can they stay here for now? They need to be implemented by a plugin which ideally would not depend on pinot-core.

**Pull request merged**
